src/net_processing.cpp

```diff
@@ -1252,6 +1253,47 @@ bool PeerManagerImpl::BlockRequestAllowed(const CBlockIndex* pindex)
            (GetBlockProofEquivalentTime(*pindexBestHeader, *pindex, *pindexBestHeader, m_chainparams.GetConsensus()) < STALE_RELAY_AGE_LIMIT);
 }
 
+bool PeerManagerImpl::FetchBlock(NodeId id, const CBlockIndex& index)
+{
+    PeerRef peer = GetPeerRef(id);
+    if (peer == nullptr) return false;
+    if (fImporting && !fReindex) return false;
+
+    // Ignore pre-segwit peers
+    LOCK(cs_main);
+    CNodeState* state = State(id);
+    if (state == nullptr) return false;
+    if (!state->fHaveWitness) return false;
+
+    // Ensure the peer has blocks at the height we need
+    const int node_sync_height = state->pindexBestKnownBlock ? state->pindexBestKnownBlock->nHeight : -1;
+    if (index.nHeight > node_sync_height) return false;
+
+    // Construct message to request the block
+    std::vector<CInv> invs{CInv(MSG_BLOCK | MSG_WITNESS_FLAG, index.GetBlockHash())};
+
+    // Mark block as in-flight unless it already is
+    if (!MarkBlockAsInFlight(id, index.GetBlockHash(), &index)) return false;
+
+    // Send block request message to the peer
+    bool success = m_connman.ForNode(id, [this, invs](CNode* pnode) {
```

Comment:
Prefer to capture `invs` by reference than making a copy:

```suggestion
    bool success = m_connman.ForNode(id, [this, &invs](CNode* pnode) {
```

Also maybe rename pnode to node if you're touching this part again.